File: algorithms/cpp/111.二叉树的最小深度.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
/*
 * @lc app=leetcode.cn id=111 lang=cpp
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
 public:
  int minDepth(TreeNode *root) {
    int level = 0;
    if (root == nullptr) return level;
    vector<TreeNode *> first_task = {root};
    queue<vector<TreeNode *> > que;
    que.push((first_task));
    while (!que.empty()) {
      level++;
      vector<TreeNode *> cur_tasks = que.front();
      vector<TreeNode *> new_tasks;
      for (auto node : cur_tasks) {
        if (node->left == nullptr && node->right == nullptr) {
          return level;
        }
        if (node->left != nullptr) {
          new_tasks.push_back(node->left);
        }
        if (node->right != nullptr) {
          new_tasks.push_back(node->right);
        }
      }
      if (!new_tasks.empty()) que.push(new_tasks);
      que.pop();
    }
    return level;
  }
};
// @lc code=end

File: algorithms/golang/111.二叉树的最小深度.go
/*
 * @lc app=leetcode.cn id=111 lang=golang
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
/**
 * 递归 - 如果元素只有一个节点，则需要包含该节点，比如二叉树退化为链表，则返回链表的长度
 *
 */
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	if root.Left == nil && root.Right != nil {
		return minDepth(root.Right) + 1
	}
	if root.Right == nil && root.Left != nil {
		return minDepth(root.Left) + 1
	}
	min := func(a, b int) int {
		if a < b {
			return a
		} else {
			return b
		}
	}
	return min(minDepth(root.Left)+1, minDepth(root.Right)+1)
}

// @lc code=end
